Adding a CHANGELOG.md file

AliSoftware opened the pull request:
### Description

As discussed during our last Hangout, this adds a CHANGELOG file — generated using Octokit from the past GitHub releases of fastlane — to use as the official changelog instead of listing changes in GH releases manually.

### Motivation and Context

Having a CHANGELOG.md file inside the repo, instead of only relying on GH releases to list the changes, has multiple benefits:

* It makes writing the list of changes easier, as the CHANGELOG file can then typically be updated on each Pull Request incrementally, instead of having to review all the past commits and adding all the entries at once at the time of doing a release
* People can also see the pending (but unreleased yet) changes before the release takes place, knowing what to expect in the next version
* It makes it easier to search in the CHANGELOG, e.g. to know if and when (which version) a given issue has been fixed. Conversely, GitHub Releases UI (and it being paginated) makes it really difficult to search
* If we (or more likely the Danger bot) ask the contributor of a PR to add an entry in the CHANGELOG to mention their change and credit themselves, that increase the acknowledgement of the contributor having helped fastlane and have a way to make them feel recognised officially and enforce that we're thankful for their contribution.

This sentiment, which helps making the community grow, can be even enforced by adding the author of each PR as part of the CHANGELOG entry. At CocoaPods and at SwiftGen, we use the following format for our CHANGELOG entries, mentioning both the PR# and the author(s) each time:

```
* Description of the change.  
  [Author 1 Name](https://github.com/author1handle)
  [Author 2 Name](https://github.com/author2handle)
  [#1234](https://github.com/fastlane/fastlane/issues/1234)
```

---

#### Note: Script used for the generation

<details>
<summary>This initial CHANGELOG.md file has been auto-generated from the GH releases using the following octokit script</summary>

```ruby
#!/usr/bin/env ruby

require 'octokit'

begin
  token = File.read('github_access_token')
rescue
  puts "You must store your GitHub access token in the gitignored 'github_access_token' file next to ths script."
  exit 1
end

client = Octokit::Client.new(:access_token => token)
releases = client.releases('fastlane/fastlane')

puts <<-INTRO
# _fastlane_ CHANGELOG

---

When contributing to fastlane, don't forget to add an entry to this CHANGELOG.md file by copy/pasting and adapting this entry template:

```
* Describe your change here. End it with a full-stop and 2 spaces.  
  [Put Your Name Here](https://github.com/yourGitHubHandle)
  [#nn](https://github.com/fastlane/fastlane/issues/nn)
```

---

## Master

* None.  
INTRO

releases.each do |release|
  puts ''
  puts "## #{release.name}"
  puts ''
  body = release.body
    .gsub(/^[-*] (.*) \(#(\d+)\)$/) do
      # Lines with a PR or Issue appearing in parentheses
      info = client.issue('fastlane/fastlane', $2)
      author_link = "[#{info.user.login}](#{info.user.html_url})"
      issue_or_pr_link = "[##{$2}](#{info.html_url})"
      "* #{$1}.  \n  #{author_link}\n  #{issue_or_pr_link}"
    end
    .gsub(/^[-*] (.*)$/) do
      # Standardizing text (non-link) lines, even those without a PR or issue reference
      line = $1.strip
      line += '.' unless line.end_with?('.')
      "* #{line}  "
    end

  puts body
end
```
</details>

AliSoftware committed:
Initial CHANGELOG.md file, generated from past GH Release notes

AliSoftware committed:
Update Dangerfile to request a CHANGELOG entry

AliSoftware commented:
\cc @fastlane/external-contributors + @asfalcone @mfurtak @krausefx @ohwutup @mpirri

As discussed during our Hangout last week.

Note that in addition to that, we can then consider automating the release process so it handles the creation of a GH Release for us, copying the appropriate CHANGELOG entries in the GH release body upon creation (see how I do it [here](https://github.com/AliSoftware/SwiftGen/blob/master/Rakefile#L247-L253)).

That could creating improve the overall release process imho.

AliSoftware committed:
Ask people to credit themselves in YourFirstPR.md

Credit is important, recognising that fastlane is built by the community and is awesome also thanks to all of YOU and your contributions!

DanToml commented on Dangerfile:
````diff
@@ -6,9 +6,23 @@ if (github.pr_body + github.pr_title).include?("WIP")
+```markdown
+* #{pr_title}#{' '}
+  [#{pr_author}](https://github.com/#{pr_author})
+  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
````
/s/CocoaPods/fastlane/

DanToml commented on YourFirstPR.md:
```diff
@@ -69,6 +69,7 @@ When the coding is done and you’re finished testing your changes, you are read
 - Create a separate PR for each tool that you’ve worked on
 - Use a descriptive title 
 - Link the issues that are related to your PR in the body
+- Add en entry in the `CHANGELOG.md` file describing your change and crediting yourself for this awesome contribution you did
```
/s/this awesome contribution you did/the awesome contribution you made

DanToml commented on CHANGELOG.md:
```diff
@@ -0,0 +1,314 @@
+# Fastlane CHANGELOG
+
```
It's may be worth adding a `## Master` header for unreleased changes

AliSoftware replied:
Indeed. Do you usually leave this `## Master` section empty at start of a new release, or put something like `_nothing yet_` there until the first PR adds an entry?

lacostej replied:
And you want to modify the top Rakefile to use this `CHANGELOG` as input.

AliSoftware replied:
@lacostej The `Rakefle` I talk about in the PR description is one for another repo of mine (SwiftGen), not one in Fastlane yet 😉  fast lane obviously uses a Fastfile to automate its release process 😛 

So basically we'll just modify [fastlane's own `Fastfile` here](https://github.com/fastlane/fastlane/blob/master/fastlane/Fastfile#L88-L89), extracting the `description` from the `CHANGELOG.md` instead of prompting for it.
The code for extracting can indeed be inspired by my Rakefile but will have to be adapted for fastlane's needs indeed :wink:

DanToml replied:
@AliSoftware CocoaPods uses a "None" placeholder (https://github.com/CocoaPods/CocoaPods/commit/06d629bcbcc431276ab2c7f2d3681d04acc42314)

lacostej replied:
@AliSoftware yes I meant `Fastfile`. You can also change this: https://github.com/fastlane/fastlane/blob/master/fastlane/Fastfile#L112

AliSoftware replied:
Indeed :+1:

endocrimes reviewed (commented):
Couple of grammatical points, but I'm a :+1: on the addition of a changelog.

AliSoftware committed:
Addressing @dantoml 's remarks 👍

KrauseFx commented on CHANGELOG.md:
```diff
@@ -0,0 +1,318 @@
+# Fastlane CHANGELOG
```
Should be `_fastlane_ changelog`

KrauseFx replied:
Meaning lower case and italic

AliSoftware replied:
Fixed in 946293ccc7dbff55a8dbc13f3133229a425bfbc3

KrauseFx commented on CHANGELOG.md:
```diff
@@ -0,0 +1,318 @@
+
+## Master
+
+* None.  
```
so after every release we want to reset the top of the file to have `Master` `None`? Can we remove the `.` here?

AliSoftware replied:
Yes, so that section doesn't stay empty right after a release is done and looks ugly.

The `.` here is not just a full-stop, it's a full-stop followed by two spaces. That's what we use at the end of _every_ changelog entry, because Markdown interpret those specially, formatting the resulting rendered text as a line break within the same paragraph (as opposed to ignoring single CRLF used for wrapping, and as opposed to creating a new paragraph when you use multiple CRLFs)

Sure, For this special case of `None` we can definitely get rid of the full-stop + 2-spaces, as there's nothing on the next line, contrary to a typical changelog entry where it's followed by the links to the author & PR# which we want to keep on the next line but in the same paragraph. But keeping the full-stop + 2 spaces here keeps the convention everywhere, even for those "dummy" entries, and makes it consistent everywhere.

KrauseFx commented on Dangerfile:
```diff
@@ -6,9 +6,23 @@ if (github.pr_body + github.pr_title).include?("WIP")
 end
 
-# Contributors should always provide a changelog when submitting a PR
+# Contributors should always provide a summary when submitting a PR
```
We don't currently run danger unfortunately due to issues when people force-push. I don't know if that has been resolved yet 😢

AliSoftware replied:
Maybe it's time to ping Orta about it to start having Danger working again on fastlane?

mfurtak commented:
I really love that you all are owning this improvement from top to bottom :+1:

KrauseFx reviewed (changes requested):
I love the approach, I'm very much in favor of changing this, however we have to also keep the GitHub releases for multiple reasons:
- We access it from the fastlane source code to show changes in line
- People are used to check it there, so it should keep working

We have to update the Fastfile to make use of this too. Note how the `fastlane bump` lane currently dumps the changelog to the PR based on the most recent commits. We'll have to parse the Changelog.md file to use the changes in the PR body and in the GitHub releases tab to then later on how it in the fastlane output.
We also want people to mention their own name next to the change, right? If so, we should describe how to do it not only in the `Dangerfile`, but also in the `Changelog.md`. Additionally I really liked that the changelog included the `#number`, which enabled people to click it to see the actual PR and all its discussion and changes.

Definitely an exciting change, with some more things that need to be done to have it ready to be merged  🚀

AliSoftware committed:
_fastlane_ italic&lowercase title + entry template

AliSoftware commented:
Next steps to finish this PR:

* [X] Introduce the CHANGELOG file, initialising it with the past changelog entries from GH releases
* [X] Agree on a CHANGELOG format for the future releases
* [ ] Make Danger working again
* [ ] Automate the release process: Making `fastlane bump` / `fastlane release` for _fastlane_ itself automatically
  * [ ] extract `release_body` = {{ the `CHANGELOG.md` entries from the `## Master` section }}
  * [ ] Rename that `## Master` section with the version number / name to release in the `CHANGELOG.md` and commit
  * [ ] Push the tag & GH release using previously extracted `release_body`
  * [ ] Inject a brand new empty `## Master` section after the tag/release has been done, for a fresh start
* [ ] Party

^ @KrauseFx does this fit the vision you described in https://github.com/fastlane/fastlane/pull/8060#pullrequestreview-19416305 ?

---

I'll code the appropriate scripts in `fastlane/Fastfile` to amend `bump`/`release` appropriately later — maybe this week-end if I have time — or if anyone can do it before, don't hesitate to take the lead.

AliSoftware commented:
> We also want people to mention their own name next to the change, right? If so, we should describe how to do it not only in the Dangerfile, but also in the Changelog.md. Additionally I really liked that the changelog included the #number, which enabled people to click it to see the actual PR and all its discussion and changes.

That's now provided by the way of a [CHANGELOG entry template at the top of the `CHANGELOG.md`](https://github.com/fastlane/fastlane/pull/8060/commits/946293ccc7dbff55a8dbc13f3133229a425bfbc3), both showing the entry format but also how to mention the link to the issue or PR so people can still click it to see actual changes 😉

ohwutup commented on CHANGELOG.md:
````diff
@@ -0,0 +1,330 @@
+When contributing to fastlane, don't forget to add an entry to this CHANGELOG.md file by copy/pasting and adapting this entry template:
+
+```
+* Describe your change here. End it with a full-stop and 2 spaces.  
````
Curious here - what does it mean to end with a "full-stop"?

AliSoftware replied:
Not a native english speaker, but that's how I was told to call a punctuation period mark, `.`, used to end a sentence :wink:

AliSoftware replied:
@ohwutup "full-stop" is UK while "period" is US, I think

ohwutup replied:
ah... je comprends maintenant! 😄 🇫🇷

milch replied:
What is the reason for ending in 2 spaces?

AliSoftware replied:
@milch Ending a line with two spaces in Markdown will render as a line break inside the same paragraph without creating a new paragraph

https://daringfireball.net/projects/markdown/syntax#p

> A paragraph is simply one or more consecutive lines of text, separated by one or more blank lines. (A blank line is any line that looks like a blank line — a line containing nothing but spaces or tabs is considered blank.) Normal paragraphs should not be indented with spaces or tabs.
>
> The implication of the “one or more consecutive lines of text” rule is that Markdown supports “hard-wrapped” text paragraphs. This differs significantly from most other text-to-HTML formatters (including Movable Type’s “Convert Line Breaks” option) which translate every line break character in a paragraph into a `<br />` tag.
>
> **When you do want to insert a `<br />` break tag using Markdown, you end a line with two or more spaces, then type return.**

This allows the author & PR links to be in a dedicated line clearly separated from the entry text, but still attached to the same bullet point.

milch replied:
Alright, that makes sense! Never knew that this worked. This is a bit of an obscure requirement though, maybe it should be checked with Danger to make sure people follow it.

AliSoftware replied:
I'd love to have a Danger script for that. [There's a "changelog" Danger plugin](https://github.com/dblock/danger-changelog) doing such syntax checks, but afaik it's a bit too restrictive on the syntax right now. Would be worth a try though once we make Danger working again.

AliSoftware committed:
Regenerate CHANGELOG.md using new script generating new format

AliSoftware committed:
Manual fixes to CHANGELOG.md to avoid some _ to be interpreted as italics

AliSoftware commented:
Just updated my Octokit script and ran it again so that the `CHANGELOG.md` file now have our new expected format.
CHANGELOG now mentions the author and PR or issue link for each entry even for past CHANGELOG entries.

hjanuschka commented:
this is really great! thx @AliSoftware

AliSoftware committed:
Links in CHANGELOG should point to the PR, not the issue, to see actual changes

DanToml commented:
@AliSoftware There's a bunch of stuff related to changelog handling in Rainforest, if you start at https://github.com/CocoaPods/Rainforest/blob/master/Rakefile#L889-L912 and work up, you'll see how we do it there

KrauseFx commented:
@DanToml thanks for linking to that part of Rainforst. I feel like this is a great candidate for a fastlane plugin as I'm sure many iOS/Android developers have a Changelog.md too they want to parse and use in _fastlane_ 👍

KrauseFx commented:
@AliSoftware I'm not a fan of the period at the end of each line, seems a bit out of place. Can we have the 2 spaces without the dot at the end?

AliSoftware commented:
Sure, as long as the two spaces are kept for formatting 👍

AliSoftware committed:
Remove period at end of CHANGELOG entries

KrauseFx commented:
So since we don't have danger yet, this could easily be part of the tests, right? Then the build would fail if the changelog.md contains an invalid row, with an error message on how to fix it

AliSoftware committed:
Add a spec to check the format of CHANGELOG.md

(Introducing a volontary mistake in the CHANGELOG to test this failure on CircleCI)

AliSoftware commented:
> Then the build would fail if the changelog.md contains an invalid row, with an error message on how to fix it

@KrauseFx like this? https://circleci.com/gh/fastlane/fastlane/7514

AliSoftware committed:
Fixing the voluntary missing space in CHANGELOG.md

KrauseFx commented:
Yeah that looks great to me, do you think people will understand?

AliSoftware commented:
Yeah I think it'll do. We can always adapt the error message later if we see it's not clear enough to people contributing, we'll see how it goes.

FYI I thought I'd have time this weekend to finish the last step of this PR (adapting fastlane's Fastfile to finally automate the release) but got more busy than I anticipated. Won't probably have much time for it next week either, but hoping to go back at it the week after (except if someone beats me to it)

mfurtak commented:
Hi folks, does this PR have a clear path forward?

AliSoftware commented:
The only missing thing in that PR is modifying the automation of the release process (_fastlane_'s own Fastfile) to adapt it (extract the CHANGELOG from the CHANGELOG.md file instead of the commits) 

To-do list here: https://github.com/fastlane/fastlane/pull/8060#issuecomment-276494552

I still haven't had any time to continue working on it (too much personal+work+OSS stuff to do lately) so if someone feels like finishing this by just adapting the Fastfile please don't hesitate to take over.

hjanuschka commented:
@AliSoftware  you see any finishing path on this? - i still would love to have a nice changelog!

AliSoftware commented:
@hjanuschka I have to admit that I don't have time these days to even work on my own repositories, with my paid work + conference talks + house hunting 😄 so feel free to continue without me on that

AliSoftware commented:
@googlebot I signed it!

ohayon commented:
Hey everyone! We have talked about this a bit internally and have decided to close this PR for now at least. We feel like it might be asking too much to have every contributor update the changelog and could increase time from PR submission to merge if we need to request that a contributor update it. We thought that maybe it might be better, if people feel strongly about wanting this in the repo, if we generate it programmatically based on release notes. Please feel free to continue discussion on here as we are always open to changing our minds and approach. 🚀

AliSoftware commented:
I think having a CHANGELOG.md file, especially to be able to search in it:
 - be able to determine in which version of fastlane was this particular bug fixed
 - or check if this bug that a user currently encounter with fastlane version X has been fixed or talked about in the CHANGELOG in newer versions

That's my 2 main problems with the CHANGELOG only be available in GH Releases currently, that we don't have a single page with all the changes. This is especially important because
 - GitHub releases are paginated: and on multiple occasions I spent lots of time searching multiple pages back and forth to find the bullet point and changelog I was interested in, that's a terrible experience
 - fastlane is very fast paced, we have release very very often. This means that if we don't update fastlane for just 2-3 weeks, there are already a ton of releases since that last update, probably already multiple pages in GH releases tab


So I totally get that you don't want to change the way to request people to do PR and don't want to impose that change to contributors, but I'd be strongly interested in at least auto-generating the CHANGELOG from the GH release notes on each release, to at least have them auto-duplicated in a single CHANGELOG.md searchable document :wink:

ohayon commented:
@AliSoftware I totally agree that having them in a nice searchable doc would be great. I also would be interested in auto-generating one from each release's notes. Maybe this is something that we can bite off in the not too distant future.

AliSoftware commented:
Given that we already generate the content of the changes during the release lane of fastlane (as we use that content to create the GitHub Release's body), it might simply be a matter of prepending `### #{version}\n\n#{release_body}\n\n` to the existing `CHANGELOG.md` during that lane? So probably quite easy to do (at least once we have an existing `CHANGELOG.md` to start from)
